### PR TITLE
Add initial implementation for java time types schemas

### PR DESCRIPTION
**Pull request by bsless**
Please find below an initial implementation for time schemas

Status: mature draft, at testing phase
Scope: JVM, schemas, transformers, generators, JSON schema, swagger

Design:

At various points users can provide a DateTimeFormatter or pattern which will be compiled to one.

- time: contains validation code, bounds check and default formats, temporal queries and parsers
- transform: dynamically pick a user provided formatter or a default one, compile to decoder/encoder for each type.
- generator: normalize all types to long, generate and rebuild the temporal type
- json schema: provide accept methods for the formats defined by the RFC
- min/max specification is clunky but all parsing code resides in transform ns

### CODE REVIEW

**Comment by bsless** [comment]
@ikitommi I believe the MR is ready for review. I tried to cover everything I could think of, lmk if there are some gaps

**Review by ikitommi — APPROVED** [review]
Thanks for the MR, it seems to cover the mandatory aspects to be used with the JVM. Few considerations:

1. code split into support namespaces (json-schema, generator etc) is a good idea and needed if/when we map these to CLJS too - to avoid bloated js-bundle sizes. That said, when using the time-schemas just with the JVM (this MR), one has to require all support namespaces, which is extra thing to remember. Options:
   1. document this properly
   2. figure out a schema module system, e.g. move the current `malli.experimental.time` as `malli.experimental.time.core` and make the `malli.experimental.time` require all the namespaces and expose just a the `schemas` - you can depend on just the part or "get it all easily"
   3. merge all code into one ns (not a good idea IMO)
2. serialization. when the date schemas are used with jsonista / transit / edn, are the values encoded properly? just strings also with Transit I guess? This doesn't have to resolved right now, but to get first-class time over the wire, we should think about transit mappings too.
3. JS/CLJS time types. When we find a viable abstraction to this, it might reflect slightly on the JVM-types too. This in mind, it would be safer to use something like `timex` (like there is/was `javax`) as the ns for now
   1. if no changes when JS/CLJS supported version ships, just rename it back from `timex` to `time`
   2. if there would be breaking changes, it obvious for library users that `:timex/offset-time` and `:time/offset-time` are different.

I would go for `1.ii` and `3`, could be done before merging or I could do those after the merge. What do you think?

File: src/malli/experimental/time/transform.clj [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        s))))
+
+(defn ->formatter
+  [x]
+  (cond
+    (instance? DateTimeFormatter x) x
+    (instance? String x) (DateTimeFormatter/ofPattern x)
+    :else (throw (ex-info "Invalid formatter" {:formatter x :type (type x)}))))
+
+(defn safe-fn
```
[the remark this answers is not in the document]
Comment (bsless) [reply src/malli/experimental/time/transform.clj:33]: You're right, it's a holdover from initial code structure, will move to `mt/-safe`

**Comment by bsless** [comment]
Regarding the split and the addition of a bunch of namespaces, we have a few options, which I'm not a big fan of:
- keep as is: will need extensive documentation, make headache for users
- 1.ii: Will still require additional docs
- merge into respective parent namespaces with reader conditionals
- keep in separate files but use `load` like clojure.core does

Serialization:
- The schemas themselves are just keywords in the registry, so that's not an issue. That issue starts with min/max properties. Serializing them is not an issue, they'll be written as strings, but deserialization is a real issue I haven't figured out yet. In the beginning the parsing code as part of the core namespace to enable writing them as strings. Couple of ways to slice this:
  - bring parsers back and use them to parse the min/max properties
  - add rudimentary parsing logic separate from the transform ns just for this purpose
  - add tagged literals, `#time/date "2020-01-01"`

`timex`:
No problem, will clean up a bit then make the change.

Thoughts?

**Comment by bsless** [comment]
I must say that after `s/:time\//:timex\//g` it just looks silly. It's just not aesthetic

**Comment by ikitommi** [comment]
- `1.ii` - I think all the docs needed is how to add the time-schemas into registry using the `malli.experimental.time` namespace?
- `:timex`, agree, kinda like `:rolex` 🤣. would `:jvm/date-time` be better? if not, any suggestions how to handle this?

**Comment by ikitommi** [comment]
serialization, I was thinking just about values, but the schema serialization is another issue. When there is JS/CLJS supported version, support for data + parser would be good. I think we can start with the current way you proposed it.

**Comment by bsless** [comment]
regarding timex, my suggestion is to just use time

Rationale - these map to the ISO-8601 formatters: https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html

If anything, the prefix could be `iso` or `iso-time`. Barring that, just `time` seems most correct. It's a universal standard we're trying to satisfy

**Comment by ikitommi** [comment]
I agree `:time` is correct in the long run. My idea was just that we could use something else in case the schemas change (types, names, properties, formats) before CLJS/JS impl is done. Otherwise, there can be a silent breakage causing hard-to-debug problems if someone has persisted the schemas into db. With new schema name, one needs to migrate the schemas.

That said, current Temporal model doesn't seem that different:

temporal-object-model

I'm merging this now, we can post-tune this before release.

Big thanks for the PR!!

**Comment by bsless** [comment]
Awesome!
I'll prepare a docs update

**Comment by ikitommi** [comment]
About the easy includes. Maybe something like:

```clojure
(ns malli.experimental.time.module
  (:require [malli.experimental.time]
            [malli.experimental.time.generator]
            [malli.experimental.time.json-schema]
            [malli.experimental.time.transform]))

(defn schemas []
  (malli.experimental.time/-time-schemas))
```

**Comment by ikitommi** [comment]
... IMO sucks, but maybe less than the alternatives.

**Comment by bsless** [comment]
What about moving it to conditional read as part of malli.core?

**Comment by ikitommi** [comment]
that would make this non-experimental. We can do that later.

**Comment by bsless** [comment]
Would have wanted something like Emacs's advice or eval-after-load mechanism

**Comment by bsless** [comment]
Can we put it in core and gate its evaluation behind a jvm property or dynamic variable?, sort of `eval-when`?

**Comment by ikitommi** [comment]
Let's not do that yet. We can merge the time-schemas into the core namespaces when we drop the `experimental` ns part after proper test-run in projects & design verification on CLJS side. Most (if not all) the projects I've done (and seen) with malli have a custom registry already so it's easy to mount in the new time schemas.

btw, I'm still on holiday this week and mostly not reading emails, so lag in responding.

**Pull request merged**